moon/controller_excavator_round2.py

```diff
@@ -97,81 +129,215 @@ def vslam_reset_time(response):
+                                    print(self.sim_time, self.robot_name, "Hauler wants to resume")
+                                except WaitRequestedException as e:
+                                    self.send_speed_cmd(0.0, 0.0)
+                                    wait_for_hauler_requested = True
+                                    print(self.sim_time, self.robot_name, "Hauler requested to wait")
+                                except VSLAMDisabledException as e:
+                                    print(self.sim_time, self.robot_name, "VSLAM: mapping disabled, waiting")
+                                    self.send_speed_cmd(0.0, 0.0)
+                                    wait_for_mapping = True
+                                except VSLAMEnabledException as e:
+                                    print(self.sim_time, self.robot_name, "VSLAM: mapping re-enabled")
+                                    wait_for_mapping = False
+
+                                self.inException = False
+                            except VirtualBumperException as e:
+                                self.send_speed_cmd(0.0, 0.0)
+                                print(self.sim_time, self.robot_name, "Bumper")
+                                if distance(self.xyz, vol_list[ind[i]]) < 1:
+                                    return ind[i] # arrived
+                                self.inException = True
+                                try:
+                                    self.go_straight(-1) # go 1m in opposite direction
+                                    angle_diff = self.get_angle_diff(vol_list[ind[i]])
+                                    self.drive_around_rock(-math.copysign(5, angle_diff)) # assume 6m the most needed
+                                except ResumeRequestedException as e:
+                                    wait_for_hauler_requested = False
+                                    print(self.sim_time, self.robot_name, "Hauler wants to resume")
+                                except WaitRequestedException as e:
+                                    self.send_speed_cmd(0.0, 0.0)
+                                    wait_for_hauler_requested = True
+                                    print(self.sim_time, self.robot_name, "Hauler requested to wait")
+                                except VSLAMDisabledException as e:
+                                    print(self.sim_time, self.robot_name, "VSLAM: mapping disabled, waiting")
+                                    self.send_speed_cmd(0.0, 0.0)
+                                    wait_for_mapping = True
+                                except VSLAMEnabledException as e:
+                                    print(self.sim_time, self.robot_name, "VSLAM: mapping re-enabled")
+                                    wait_for_mapping = False
+
+                                self.inException = False
+                            except VSLAMDisabledException as e:
```

Comment:
this "tripple exception catch" is hard to follow ...

Reply:
If I am looking correctly, the outer one is just for BusShutdownException and doesn't really count, just a wrapper around the whole thing. Then we have exceptions (bumper, vslam down) which we try to handle somehow but there can be exception as we are handling. In a few places I queued up a different action to be fed back to the main 'try' block immediately but for other scenarios so far I found it easier to just deal with the problem within the exception block and be prepared if another exception happens during. Not married to it but there are trade-offs either way.